File: algorithms/lua/mergesort.lua
local function merge(arr, tmp, l, m, r)
    for i = l, r do tmp[i] = arr[i] end
    local i, j, k = l, m + 1, l
    while i <= m and j <= r do
        if tmp[i] <= tmp[j] then arr[k] = tmp[i]; i = i + 1
        else                     arr[k] = tmp[j]; j = j + 1 end
        k = k + 1
    end
    while i <= m do arr[k] = tmp[i]; i = i + 1; k = k + 1 end
    while j <= r do arr[k] = tmp[j]; j = j + 1; k = k + 1 end
end

local function mergesort(arr, tmp, l, r)
    if l >= r then return end
    local m = l + math.floor((r - l) / 2)
    mergesort(arr, tmp, l, m)
    mergesort(arr, tmp, m + 1, r)
    merge(arr, tmp, l, m, r)
end

local function main()
    local n   = tonumber(io.read("*l"))
    local arr = {}
    local i   = 1
    for v in io.read("*l"):gmatch("%S+") do
        arr[i] = tonumber(v); i = i + 1
    end
    local tmp = {}
    for j = 1, n do tmp[j] = 0 end

    local start = os.clock()
    mergesort(arr, tmp, 1, n)
    local ms = (os.clock() - start) * 1000

    print(table.concat(arr, " "))
    io.stderr:write(string.format("%.3f\n", ms))
end

main()

File: algorithms/rust/mergesort.rs
use std::io::{self, Read, Write, BufWriter};
use std::time::Instant;

fn merge(arr: &mut [i64], tmp: &mut [i64], l: usize, m: usize, r: usize) {
    tmp[l..=r].copy_from_slice(&arr[l..=r]);
    let (mut i, mut j, mut k) = (l, m + 1, l);
    while i <= m && j <= r {
        if tmp[i] <= tmp[j] { arr[k] = tmp[i]; i += 1; }
        else                 { arr[k] = tmp[j]; j += 1; }
        k += 1;
    }
    while i <= m { arr[k] = tmp[i]; i += 1; k += 1; }
    while j <= r { arr[k] = tmp[j]; j += 1; k += 1; }
}

fn mergesort(arr: &mut [i64], tmp: &mut [i64], l: usize, r: usize) {
    if l >= r { return; }
    let m = l + (r - l) / 2;
    mergesort(arr, tmp, l, m);
    mergesort(arr, tmp, m + 1, r);
    merge(arr, tmp, l, m, r);
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_ascii_whitespace();

    let n: usize = iter.next().unwrap().parse().unwrap();
    let mut arr: Vec<i64> = (0..n)
        .map(|_| iter.next().unwrap().parse().unwrap())
        .collect();
    let mut tmp = vec![0i64; n];

    let start = Instant::now();
    if n > 1 { mergesort(&mut arr, &mut tmp, 0, n - 1); }
    let ms = start.elapsed().as_secs_f64() * 1000.0;

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    for (i, v) in arr.iter().enumerate() {
        if i > 0 { write!(out, " ").unwrap(); }
        write!(out, "{}", v).unwrap();
    }
    writeln!(out).unwrap();
    eprintln!("{:.3}", ms);
}